lib/rpmchecksig.c

```diff
@@ -169,8 +169,8 @@ rpmRC rpmpkgRead(struct rpmvs_s *vs, FD_t fd,
 	goto exit;
     }
 
-    /* Read the signature header. Might not be in a contiguous region. */
-    if (hdrblobRead(fd, 1, 0, RPMTAG_HEADERSIGNATURES, sigblob, &msg))
+    /* Read the signature header. Must be in a contiguous region. */
+    if (hdrblobRead(fd, 1, 1, RPMTAG_HEADERSIGNATURES, sigblob, &msg))
```

Comment:
So, we deliberately changed this in https://github.com/rpm-software-management/rpm/commit/34c2ba3c6a80a778cdf2e42a9193b3264e08e1b3 to make it compatible with third-party RPMs that are not crafted by `rpmbuild`. I don't know if we want to revert this again...

Reply:
Personally, I would prefer to revert https://github.com/rpm-software-management/rpm/commit/34c2ba3c6a80a778cdf2e42a9193b3264e08e1b3.  The reason is attack surface: we run quite a bit of code for non-contiguous headers that we don’t run for contiguous headers.  Since this code is run before signature verification, this is significant.  I would rather not increase our attack surface to work around broken third-party signers.

Reply:
That said, I also recommend that non-contiguous header support be controlled by a flag (off by default

Reply:
NAK on this, there's an entirely reasonable case for non-contiguous signature header that we might actually adopt. Please do check back on history when suggesting changes, there's often a reason things are the way they are.

Reply:
That’s understandable, so long as we never call `headerImport` on this header.

Reply:
It turns out we do import the signature header in at least some cases.  Since the signature header is not itself signed, I would like to make sure that its parsing is as simple as possible, to reduce attack surface.  @pmatilai do we need the full generality of a non-contiguous header, or can we require that there are no duplicate tags in the signature header?  If there were guaranteed to be no duplicate tags in the signature header, I would be far more confident.

Reply:
To elaborate: most of the complicated code ― and, in particular, most of the writes that could (if there is a bug) go out of bounds ― are in the code that handles duplicate header tags.  However, I can’t see any legitimate reasons for a package to have a duplicate header tag.  In particular, I believe the set of tags in signatures is disjoint with the set of tags in immutable headers.  Therefore, it seems to me that a package with duplicate header tags is broken and should be rejected.

That said, I am fine with reverting this for now.

Reply:
(For the record: I am thinking *purely* from a security perspective here.  Of course, security is not the only goal of the RPM project.)

Reply:
Closing this.  It turns out that the relevant code is at least somewhat robust (it passed fuzzing with libfuzzer for a couple of hours with no bugs found).  I still recommend removing it from the attack surface, but for now there are bigger wins to be had elsewhere.

Reply:
You seriously think you're the first ever person to throw a fuzzer at rpm? Really now...